File: cpp/PhotonLibOS/fib.cpp
// An implementation of the recursive fork fibonacci parallelism test.
//
// Canonical (TooManyCooks) implementation: ../TooManyCooks/fib.cpp
//
// A fork spawns fib(n-1) as a bench::thread (created on the current
// vcpu and migrated to a work-pool vcpu) and the caller continues with
// fib(n-2) on its own stack, then joins the child - the same
// fork-continue-join shape as the Go and TMC implementations.

#include "memusage.hpp"
#include "photon_bench.hpp"

#include <chrono>
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

static const std::size_t iter_count = 1;

static std::uint64_t fib(std::uint64_t n) {
  if (n < 2) {
    return n;
  }
  std::uint64_t x;
  bench::thread child([&x, n] { x = fib(n - 1); });
  const std::uint64_t y = fib(n - 2);
  child.join();
  return x + y;
}

int main(int argc, char* argv[]) {
  if (argc < 2) {
    std::printf("Usage: fib <n-th fibonacci number requested>\n");
    return 0;
  }
  const std::uint64_t n = strtoull(argv[1], nullptr, 10);
  std::size_t thread_count = bench::default_thread_count();
  if (argc > 2) {
    thread_count = static_cast<std::size_t>(atoi(argv[2]));
  }
  std::printf("threads: %zu\n", thread_count);

  bench::quiet_logs();
  bench::use_pooled_stacks();

  bench::run_in_pool(thread_count, [n] {
    (void)fib(30); // warmup

    auto startTime = std::chrono::high_resolution_clock::now();
    std::uint64_t result = 0;
    for (std::size_t i = 0; i < iter_count; ++i) {
      result = fib(n);
    }
    auto endTime = std::chrono::high_resolution_clock::now();
    auto totalTimeUs = std::chrono::duration_cast<std::chrono::microseconds>(
      endTime - startTime
    );
    std::printf("output: %" PRIu64 "\n", result);
    std::printf("runs:\n");
    std::printf("  - iteration_count: %zu\n", iter_count);
    std::printf("    duration: %" PRIu64 " us\n",
                static_cast<std::uint64_t>(totalTimeUs.count()));
    std::printf("    max_rss: %ld KiB\n", peak_memory_usage());
  });
}

File: rust/forte/src/bin/fib.rs
// An implementation of the recursive fork fibonacci parallelism test.
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/fib.cpp
//
// forte is a heartbeat-scheduled fork-join scheduler: its `Worker::join` is
// almost identical to rayon's `join`, but forking is nearly free and work is
// only actually shared with another worker on a periodic heartbeat (the same
// design as chili/spice). Each recursive call threads the `&Worker` down so the
// join tree stays anchored to one pool.

use forte::Worker;
use forte_bench::{default_threads, peak_memory_usage, size_pool, POOL};
use std::time::Instant;

const ITER_COUNT: usize = 1;

fn fib(n: u64, w: &Worker) -> u64 {
    if n < 2 {
        return n;
    }
    let (x, y) = w.join(|w| fib(n - 1, w), |w| fib(n - 2, w));
    x + y
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        println!("Usage: fib <n-th fibonacci number requested>");
        return;
    }
    let n: u64 = args[1].parse().expect("n");
    let thread_count = if args.len() > 2 {
        args[2].parse().expect("thread count")
    } else {
        default_threads()
    };

    size_pool(thread_count);
    println!("threads: {}", thread_count);

    // The whole workload runs inside `with_worker` so the calling thread joins
    // the pool as a worker (analogous to rayon's `pool.install`).
    POOL.with_worker(|w| {
        let _ = fib(30, w); // warmup

        let start = Instant::now();
        let mut result = 0u64;
        for _ in 0..ITER_COUNT {
            result = fib(n, w);
        }
        let dur = start.elapsed();
        println!("output: {}", result);

        println!("runs:");
        println!("  - iteration_count: {}", ITER_COUNT);
        println!("    duration: {} us", dur.as_micros());
        println!("    max_rss: {} KiB", peak_memory_usage());
    });
}
